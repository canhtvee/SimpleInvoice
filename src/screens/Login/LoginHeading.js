import React from 'react';
import {AppText} from '../../commons';
import {Sizes} from '../../utils';

export function LoginHeading() {
  return (
    <AppText
      style={{
        alignSelf: 'center',
        fontSize: Sizes.headline,
        fontWeight: 'bold',
        marginTop: Sizes.wpx(60),
      }}>
      Login to use Simple Invoice
    </AppText>
  );
}
