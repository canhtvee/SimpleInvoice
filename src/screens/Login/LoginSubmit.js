import React from 'react';
import {Alert} from 'react-native';
import {useFormState} from 'react-hook-form';

import {AppButton} from '../../commons';
import {AppAccount, FetchApi, resetToHome, Sizes} from '../../utils';

export function LoginSubmit({control, handleSubmit}) {
  const {isSubmitting} = useFormState({control});

  const onSubmit = async data => {
    try {
      const loginResult = await FetchApi.login(data);
      if (!loginResult?.access_token) {
        throw new Error('loginResult error');
      }

      AppAccount.set(loginResult);

      //Get user profile right after sucessfully login
      const userProfile = await FetchApi.getUserProfile();
      if (!userProfile?.data) {
        throw new Error('userProfile error');
      }

      AppAccount.set({
        ...userProfile.data,
        org_token: userProfile.data.memberships[0].token,
      });

      resetToHome();
    } catch (error) {
      AppAccount.set();
      Alert.alert(null, 'Login failed \nPlease try again!');
    }
  };

  return (
    <AppButton
      title="Login"
      onPress={handleSubmit(onSubmit)}
      isLoading={isSubmitting}
      style={{marginTop: Sizes.wpx(40)}}
    />
  );
}
