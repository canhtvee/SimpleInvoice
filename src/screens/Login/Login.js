import React from 'react';
import {useForm} from 'react-hook-form';
import {AppContainer, AppTextInput} from '../../commons';
import {Sizes} from '../../utils';
import {LoginHeading} from './LoginHeading';
import {LoginSubmit} from './LoginSubmit';

export function Login() {
  const {control, handleSubmit} = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    defaultValues: {
      clientId: 'oO8BMTesSg9Vl3_jAyKpbOd2fIEa',
      clientSecret: '0Exp4dwqmpON_ezyhfm0o_Xkowka',
    },
  });

  return (
    <AppContainer style={{paddingHorizontal: Sizes.wpx(30)}}>
      <LoginHeading />
      <AppTextInput
        label="Client ID"
        control={control}
        name="clientId"
        rules={{required: {value: true, message: 'ID is required'}}}
        placeholder={'Enter ClientId'}
        containerStyle={{marginTop: Sizes.wpx(30)}}
      />
      <AppTextInput
        label="Client Secret"
        control={control}
        name="clientSecret"
        rules={{required: {value: true, message: 'Secret is required'}}}
        placeholder={'Enter Client Secret'}
        secureTextEntry
        containerStyle={{marginTop: Sizes.wpx(20)}}
      />
      <LoginSubmit control={control} handleSubmit={handleSubmit} />
    </AppContainer>
  );
}
