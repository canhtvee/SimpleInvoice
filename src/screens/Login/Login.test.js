import React from 'react';
import {act, cleanup, fireEvent, render} from '@testing-library/react-native';

import {Login} from './Login';

/**
 * Run test command
 * yarn test -u -t="Login"
 */

afterEach(cleanup);

const mockLogin = jest.fn();
const mockFetchUser = jest.fn();

jest.mock('../../utils/fetchApi/FetchApi', () => ({
  FetchApi: {
    login: data => mockLogin(data),
    getUserProfile: () => mockFetchUser(),
  },
}));

const getElement = tree => {
  const clientId = tree.getByPlaceholderText('Enter ClientId');
  const clientSecret = tree.getByPlaceholderText('Enter Client Secret');
  const loginSubmit = tree.getByText('Login');

  return {clientId, clientSecret, loginSubmit};
};

describe('Login', () => {
  test('should render correctly', () => {
    const tree = render(<Login />);

    expect(tree).toMatchSnapshot();

    const {clientId, clientSecret} = getElement(tree);

    expect(clientId.props.value).toBe('oO8BMTesSg9Vl3_jAyKpbOd2fIEa');
    expect(clientSecret.props.value).toBe('0Exp4dwqmpON_ezyhfm0o_Xkowka');
  });

  test('should not login if input empty', async () => {
    const tree = render(<Login />);

    const {clientId, clientSecret, loginSubmit} = getElement(tree);

    await act(async () => {
      fireEvent.changeText(clientId, '');
      fireEvent.changeText(clientSecret, '');
      fireEvent.press(loginSubmit);
    });

    expect(mockLogin).toHaveBeenCalledTimes(0);
  });

  test('should login if input valid', async () => {
    const tree = render(<Login />);

    const {clientId, clientSecret, loginSubmit} = getElement(tree);

    await act(async () => {
      fireEvent.changeText(clientId, 'test');
      fireEvent.changeText(clientSecret, 'test');
      fireEvent.press(loginSubmit);
    });

    expect(mockLogin).toBeCalledWith({
      clientId: 'test',
      clientSecret: 'test',
    });
  });
});
