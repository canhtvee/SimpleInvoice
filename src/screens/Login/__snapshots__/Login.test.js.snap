// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    [
      "left",
      "right",
    ]
  }
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "flex-start",
      },
      {
        "paddingHorizontal": 59,
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 27.5,
        },
        {
          "alignSelf": "center",
          "fontSize": 43.5,
          "fontWeight": "bold",
          "marginTop": 118.5,
        },
      ]
    }
  >
    Login to use Simple Invoice
  </Text>
  <View
    style={
      {
        "marginTop": 59,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontSize": 27.5,
          },
          undefined,
        ]
      }
    >
      Client ID
    </Text>
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#0000008A",
            "borderRadius": 6,
            "borderWidth": 1,
            "flexDirection": "row",
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Enter ClientId"
        spellCheck={false}
        style={
          [
            {
              "flex": 1,
              "fontSize": 27.5,
              "paddingHorizontal": 16,
            },
            {
              "paddingHorizontal": 16,
              "paddingVertical": 23.5,
            },
            {
              "color": "#000000",
            },
            undefined,
          ]
        }
        value="oO8BMTesSg9Vl3_jAyKpbOd2fIEa"
      />
    </View>
  </View>
  <View
    style={
      {
        "marginTop": 39.5,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontSize": 27.5,
          },
          undefined,
        ]
      }
    >
      Client Secret
    </Text>
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#0000008A",
            "borderRadius": 6,
            "borderWidth": 1,
            "flexDirection": "row",
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Enter Client Secret"
        secureTextEntry={true}
        spellCheck={false}
        style={
          [
            {
              "flex": 1,
              "fontSize": 27.5,
              "paddingHorizontal": 16,
            },
            {
              "paddingHorizontal": 16,
              "paddingVertical": 23.5,
            },
            {
              "color": "#000000",
            },
            undefined,
          ]
        }
        value="0Exp4dwqmpON_ezyhfm0o_Xkowka"
      />
      <Text
        allowFontScaling={false}
        onPress={[Function]}
        selectable={false}
        style={
          [
            {
              "color": "#000000DE",
              "fontSize": 35.5,
            },
            {
              "paddingRight": 16,
            },
            {
              "fontFamily": "Feather",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        ï…£
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={16}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2196f3",
        "borderRadius": 6,
        "height": 79,
        "justifyContent": "center",
        "marginTop": 79,
        "opacity": 1,
        "paddingHorizontal": 47,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 29.5,
            "fontWeight": "600",
          },
          {
            "color": "#FFFFFF",
          },
          undefined,
          undefined,
        ]
      }
    >
      Login
    </Text>
  </View>
</RNCSafeAreaView>
`;
